### PR TITLE
chain/ethereum,graph,node: start backfill BlockIngestor process for unprocessed blocks

### PR DESCRIPTION
**Pull request by colindickson**
Since the Firehose Block Ingestor starts at the current HEAD block of the chain, the purpose of this change is to spawn a process which will backfill the blocks from the start of the chain all the way to the block where the ingestor first started processing its blocks.

### CODE REVIEW

File: graph/src/blockchain/firehose_block_ingestor.rs [hunk 1]
```diff
@@ -82,6 +82,87 @@ where
...
+        let mut backfill_cursor = self.fetch_backfill_cursor().await;
+
+        loop {
+            let backfill_completed = self.fetch_backfill_is_completed().await;
+            if backfill_completed {
+                info!(self.logger, "Backfill completed");
+                break;
+            }
+
+            let backfill_target_block_number = self.fetch_backfill_target_block_num().await;
```
Comment (maoueh) [inline graph/src/blockchain/firehose_block_ingestor.rs:100]: This should be done outside the main loop, there is no need to fetch the backfill target block num on every loop, only once.

File: graph/src/blockchain/firehose_block_ingestor.rs [hunk 2]
```diff
@@ -82,6 +82,87 @@ where
...
+            // If we reach this point, we must wait a bit before retrying
+            backoff.sleep_async().await;
+        }
+    }
+
+    async fn fetch_backfill_is_completed(&self) -> bool {
+        let mut backoff =
+            ExponentialBackoff::new(Duration::from_millis(250), Duration::from_secs(30));
+        loop {
+            match self.chain_store.clone().chain_backfill_is_completed() {
```
Comment (maoueh) [inline graph/src/blockchain/firehose_block_ingestor.rs:140]: I feel `chain_backfill_is_completed` should return `Result<bool, Error>`. What the case where it can be `Ok(None)` exactly? Wouldn't it better to return in `Error` if backfill completeness can't be determined somehow?

File: graph/src/blockchain/firehose_block_ingestor.rs (hunk 1 above)
Comment (maoueh) [inline graph/src/blockchain/firehose_block_ingestor.rs:100]: I would like to have `fetch_backfill_target_block_num` split in two concerns, retrieve the target backfill block num alone and determine/set the backfill target block num if not set. Pseudo-code here would look like:
```suggestion
            let mut backfill_target_block_number = self.fetch_backfill_target_block_num().await;
            if backfill_target_block_number.is_none() {
                backfill_target_block_number =  Some(self.initialize_backfill_target_block_num().await);
            }
```

File: graph/src/blockchain/firehose_block_ingestor.rs [hunk 3]
```diff
@@ -97,6 +178,99 @@ where
...
+                }
+                Err(e) => {
+                    error!(self.logger, "Setting chain backfill target failed: {:?}", e);
+                    backoff.sleep_async().await;
+                }
+            }
+        }
+    }
+
+    async fn set_backfill_target_block_num(&self, block_num: BlockNumber) {
```
Comment (maoueh) [inline graph/src/blockchain/firehose_block_ingestor.rs:254]: When splitting `fetch_backfill_target_block_num`, I don't think this will be necessary and could be put straight in `initialize_backfill_target_block_num` which would make the retry portion clearer and located in the `initialize_backfill_target_block_num` directly.

File: graph/src/blockchain/firehose_block_ingestor.rs [hunk 4]
```diff
@@ -159,10 +333,87 @@ where
 
         self.chain_store
             .clone()
-            .set_chain_head(block, response.cursor.clone())
+            .set_chain_head(block.clone(), response.cursor.clone())
```
Comment (maoueh) [inline graph/src/blockchain/firehose_block_ingestor.rs:336]: Why is this required now? Isn't this in untouched portion of your PR?

File: graph/src/blockchain/firehose_block_ingestor.rs [hunk 5]
```diff
@@ -159,10 +333,87 @@ where
...
+            .set_chain_backfill_cursor(response.cursor.clone())
+            .await
+            .context("Updating chain backfill cursor")?;
+
+        if block_number >= backfill_target_block_number.into() {
+            self.chain_store
+                .clone()
+                .set_chain_backfill_as_completed()
+                .await
+                .context("setting backfill completion")?;
```
Comment (maoueh) [inline graph/src/blockchain/firehose_block_ingestor.rs:414]:
```suggestion
                .context("Setting backfill completion")?;
```

**Review by maoueh — CHANGES_REQUESTED** [review]

File: graph/src/blockchain/firehose_block_ingestor.rs (hunk 1 above)
Comment (colindickson) [reply graph/src/blockchain/firehose_block_ingestor.rs]: done!

File: graph/src/blockchain/firehose_block_ingestor.rs (hunk 3 above)
Comment (colindickson) [reply graph/src/blockchain/firehose_block_ingestor.rs]: done!

File: graph/src/blockchain/firehose_block_ingestor.rs (hunk 2 above)
Comment (colindickson) [reply graph/src/blockchain/firehose_block_ingestor.rs:144]: done!

**Comment by neysofu** [comment]
@maoueh should I close this?

**Comment by maoueh** [comment]
I think it might not be important anymore. It was a way to backfill missing block from `FirehoseBlockIngestor` that start at "live" point of the chain.

Nobody ever reported any problem around such missing feature, so I imagine it's not needed.

**Comment by lutter** [comment]
Closing this since it's not been active for a while. Feel free to reopen when you're ready to work on this again
